pkg/controller/hibernation/aws_actuator_test.go

```diff
@@ -118,6 +125,62 @@ func TestStopAndStartMachines(t *testing.T) {
 	}
 }
 
+func TestStopPreemptibleMachines(t *testing.T) {
+	tests := []struct {
+		name string
+		// number of on-demand and spot instances
+		onDemand int
+		spot     int
+
+		setupClient func(*testing.T, *mockawsclient.MockClient)
+	}{{
+		name:     "only on-demand instances",
+		onDemand: 1,
+		setupClient: func(t *testing.T, c *mockawsclient.MockClient) {
+			c.EXPECT().StopInstances(gomock.Any()).Do(
+				func(input *ec2.StopInstancesInput) {
+					input.InstanceIds = aws.StringSlice([]string{"onDemand-0"})
+				}).Return(nil, nil)
+		},
+	}, {
+		name: "only spot instances",
+		spot: 1,
+		setupClient: func(t *testing.T, c *mockawsclient.MockClient) {
+			c.EXPECT().TerminateInstances(gomock.Any()).Do(
+				func(input *ec2.TerminateInstancesInput) {
+					input.InstanceIds = aws.StringSlice([]string{"spot-0"})
+				}).Return(nil, nil)
+		},
+	}, {
+		name:     "mix of spot and on-demand instances",
+		onDemand: 1,
+		spot:     2,
+		setupClient: func(t *testing.T, c *mockawsclient.MockClient) {
+			c.EXPECT().StopInstances(gomock.Any()).Do(
+				func(input *ec2.StopInstancesInput) {
+					input.InstanceIds = aws.StringSlice([]string{"onDemand-0"})
+				}).Return(nil, nil)
+			c.EXPECT().TerminateInstances(gomock.Any()).Do(
+				func(input *ec2.TerminateInstancesInput) {
+					input.InstanceIds = aws.StringSlice([]string{"spot-0", "spot-1"})
+				}).Return(nil, nil)
```

Comment:
At least for this test case, it would be nice to assert the number of calls to each method.
```suggestion
			c.EXPECT().StopInstances(gomock.Any()).Times(1).Do(
				func(input *ec2.StopInstancesInput) {
					input.InstanceIds = aws.StringSlice([]string{"onDemand-0"})
				}).Return(nil, nil)
			c.EXPECT().TerminateInstances(gomock.Any()).Times(2).Do(
				func(input *ec2.TerminateInstancesInput) {
					input.InstanceIds = aws.StringSlice([]string{"spot-0", "spot-1"})
				}).Return(nil, nil)
```

Reply:
there are only 2 calls at max , one call for all on-demand, and another for all spot instances.

Reply:
orite. I wonder why the above change didn't fail for me?

Reply:
Maybe it didn't fail b/c there is no `defer ctrl.Finish()` after the `gomock.NewController(t)` line???

Reply:
Goooood call.